### PR TITLE
LegendIcon: only render color picker when the props change

### PR DESCRIPTION
**Pull request by ryantxu**
The current `React.FunctionComponent<Props>` re-renders the color picker with each data update, this put a simple component around them that only updates when color/disabled change

### CODE REVIEW

File: packages/grafana-ui/src/components/VizLegend/VizLegendSeriesIcon.tsx [hunk 1]
```diff
@@ -11,22 +11,31 @@ interface Props {
...
-      )}
-    </SeriesColorPicker>
-  );
-};
+export class VizLegendSeriesIcon extends Component<Props> {
+  shouldComponentUpdate(nextProps: Props) {
+    if (this.props.color !== nextProps.color) {
+      return true;
+    }
+    return this.props.disabled !== nextProps.disabled;
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/VizLegend/VizLegendSeriesIcon.tsx:19]:
```suggestion
return this.props.color !== nextProps.color || this.props.disabled !== nextProps.disabled;
```

**Comment by leeoniya** [comment]
small suggestion, otherwise LGTM.

**Review by leeoniya — APPROVED** [review]

**Comment by hugohaggmark** [comment]
Couldn't we use React.memo around the whole component instead of turning it into a class component?

**Comment by ryantxu** [comment]
> Couldn't we use React.memo around the whole component instead of turning it into a class component?

I tried, but it still renders every time -- it needs to ignore changes to `onColorChange`, (that gets a new instance every time)

Is React.memo better/different than a component?

**Comment by hugohaggmark** [comment]
> > Couldn't we use React.memo around the whole component instead of turning it into a class component?
> 
> I tried, but it still renders every time -- it needs to ignore changes to `onColorChange`, (that gets a new instance every time)
> 
> Is React.memo better/different than a component?

Gotcha, have you tried converting this
https://github.com/grafana/grafana/blob/5f54f2dc000fb49bd71af28845dce6060dcc0df9/packages/grafana-ui/src/components/VizLegend/VizLegendListItem.tsx#L27

and this
https://github.com/grafana/grafana/blob/5f54f2dc000fb49bd71af28845dce6060dcc0df9/packages/grafana-ui/src/components/VizLegend/VizLegendTableItem.tsx#L35

from anonymous functions to memoized functions?

**Comment by Clarity-89** [comment]
React.memo has accepts a [second argument](https://reactjs.org/docs/react-api.html#reactmemo), which is a custom comparison function. In this case it's probably gonna look similar to `shouldComponentUpdate`, but used with functional component.

**Pull request merged**